oktanaut blog (#1470)
oktadeveloper/okta-blog: _source/_posts/2023-10-12-okta-documentation-chatbot.md

tags: [python-chatbot, python, okta-documentation, ai, automation]
tweets:
- ""
image: blog/okta-documentation-chatbot/oktanaut.jpeg
type: awareness
github: https://github.com/oktadev/okta-python-chatbot-example
---

In today's fast-paced world of technology, developers must navigate through detailed API documentation to integrate with software solutions. As a Developer Support Intern at Okta, I noticed developers underutilizing [Okta's Developer Documentation] (developer.okta.com). To help them benefit from the wealth of information in these docs, I was inspired to build a tool using OpenAI. Leveraging ChatGPT, I built ✨Oktanaut✨, a versatile Python chatbot running on Jupyter Notebook that makes access to information on Okta's Developer Documentation simple and straightforward.

{% img blog/okta-documentation-chatbot/oktadevforumsdk.jpeg alt:"OktaDev Forum and SDK logo" height:"400" %}{: .center-image } 
– Image source: <cite>[Okta Developer GitHub](https://raw.githubusercontent.com/oktadev/.github/main/images/okta-dev-header.png)</cite> 

## Evolution of Oktanaut, a Python Chatbot 

> I developed two versions of Oktanaut. The first version was a specific approach, offering greater precision rather than general knowledge. It built this chatbot using LlamaIndex and trained it on the Okta Developer Documentation. While this version generated more accurate responses due to its focused training, it could not handle diverse questions.

Reviewer: ```suggestion
I developed two versions of Oktanaut. The first version was a specific approach, offering greater precision rather than general knowledge. I built the prototype chatbot using LlamaIndex and trained it on the Okta Developer Documentation. While this version generated accurate responses due to its focused training, it could not handle diverse questions.
```

> The second version, a versatile approach, utilized the capabilities of the [OpenAI GPT-3.5 API](https://chat.openai.com/auth/login); it generated responses based on a broader sample of user inquiries. This version handled various questions, providing a dynamic, self-learning, and interactive experience.

Reviewer: ```suggestion
The second version, a versatile approach, utilized the capabilities of the [OpenAI GPT-3.5 API](https://chat.openai.com/auth/login). It generated responses based on a broader sample of user inquiries. This version handled various questions, providing a dynamic, self-learning, and interactive experience.
```

The trade-off between precision and versatility was obvious when comparing the two versions. The LLamaIndex-based Oktanaut version excelled in accuracy, while the GPT-3.5-powered version generated better responses to a broader set of questions. The final chatbot I chose to demo in this blog uses OpenAI API as it has bigger models to run inquiries and is more user-friendly. 

## Meticulous Training on Okta's Developer Documentation for Enhanced Performance

> For both versions, meticulous training was a necessary building block. Doing so allowed for more accurate responses to edge cases and custom-specific questions. I collected training data, including curated sample questions, answers, and information provided by Okta's Developer Support Engineers. I stored this data in Google Drive to run on [Colab](https://colab.google/). Additionally, I transformed the chatbot into a self-learning entity by storing historical interactions, allowing it to learn from its mistakes. This self-improvement mechanism allowed the chatbot to utilize context information from past responses within the session to provide more informed responses.

Reviewer: ```suggestion
For both versions, meticulous training was a necessary building block. Carefully refining the training data helped the bot produce more accurate responses to edge cases and custom-specific questions. I collected training data, including curated sample questions, answers, and information provided by Okta's Developer Support Engineers. I stored this data in Google Drive to run on [Colab](https://colab.google/). Additionally, I transformed the chatbot into a self-learning entity by storing historical interactions, allowing it to learn from its mistakes. This self-improvement mechanism allowed the chatbot to utilize context information from past responses within the session to provide more informed responses.
```

## Personalizing the Conversational Experience with the Python Chatbot AI
I programmed the chatbot to greet users with its name, functionality, and purpose. Doing so added a personal touch that will help minimize potential confusion users may have when interacting with the chatbot.

> I utilized a Python library called [Panel](https://panel.holoviz.org/getting_started/installation.html) on the frontend to achieve a user-friendly interface. Building the interface was not simple as 1)Panel library has several bugs and glitches, and 2)considering a frontend built-in Python is uncommon. Also, another thing to note, when I considered deployment options, I decided to run Oktanaut on Jupyter Notebook to ensure a simple setup process for everyone.

Reviewer: ```suggestion
I utilized a Python library called [Panel](https://panel.holoviz.org/getting_started/installation.html) on the frontend to achieve a user-friendly interface. Building the interface was not as simple as I first expected. The Panel library has several bugs and glitches, and considering a frontend built-in Python is uncommon. When I considered deployment options, I decided to run Oktanaut on Jupyter Notebook to ensure a simple setup process for everyone.
```

Reviewer: idk what "considering a frontend built-in Python is uncommon" means but you can leave it

 
## Demonstrating Oktanaut's AI Chat Potential

> The completed chatbot script can be found [here](https://github.com/oktadev/okta-python-chatbot-example).  

Reviewer: ```suggestion
The completed code for the OpenAI version of the chatbot can be found [here](https://github.com/oktadev/okta-python-chatbot-example).  
```

def add_prompts_conversation(_):
    prompt = client_prompt.value_input
    client_prompt.value = ''

    context.append({'role':'user', 'content':f"{prompt}"})

    response = continue_conversation(context)

    context.append({'role':'assistant', 'content':f"{response}"})

    panels.append(
        pn.Row('Developer:', pn.pane.Markdown(prompt, width=600)))
    panels.append(
        pn.Row('Oktanaut:', pn.pane.Markdown(response, width=600)))

    return pn.Column(*panels)
```

```py
context = [ {'role':'system', 'content':"""'I am Oktanaut, a helpful chatbot meant to answer your questions about Okta and OAuth Developer Documentation. I answer questions from Okta Developer Documentation. How may I help you? (Say "thank you" to end the session) \n'"""} ]

pn.extension()

panels = []

client_prompt = pn.widgets.TextInput(value="Hi", placeholder='Enter your questions here...')
button_conversation = pn.widgets.Button(name="Chat with Oktanaut!")

interactive_conversation = pn.bind(add_prompts_conversation, button_conversation)

dashboard = pn.Column(
    client_prompt,
    pn.Row(button_conversation),
    pn.panel(interactive_conversation, loading_indicator=True),
)

dashboard
```

> The code for the Jupiter Python Notebook above [okta-python-chatbot-example

Reviewer: ```suggestion
The code for the Jupiter Python Notebook above, [okta-python-chatbot-example
```

> /GPTChatbot.ipynb](https://github.com/oktadev/okta-python-chatbot-example/blob/main/GPTChatbot.ipynb) is for summoning Oktanaut. 

Reviewer: ```suggestion
/GPTChatbot.ipynb](https://github.com/oktadev/okta-python-chatbot-example/blob/main/GPTChatbot.ipynb), is for summoning Oktanaut. 
```

Here's a step-by-step walkthrough of how to run the code:

1. With a Google account, ensure you have access to [Colab](https://colab.google/); this will be the environment we will use to run the Python scripts.
 
2. Library installation: The script begins with installing the required Python packages: 
- [OpenAI](https://pypi.org/project/openai/) 
- [Panel](https://panel.holoviz.org/getting_started/installation.html)
- [LlamaIndex](https://pypi.org/project/llama-index/)

> These packages work with the GPT-3.5 model, create the web interface, and manage the chatbot's conversation. Note, LlamaIndex was used only to read the data.

Reviewer: ```suggestion
These packages work with the GPT-3.5 model, create the web interface, and manage the chatbot's conversation. The LlamaIndex package is only used to read data in the OpenAI implementation of the bot. 
```

> 3. Create an OpenAI account and follow these [instructions](https://help.openai.com/en/articles/4936850-where-do-i-find-my-api-key) to obtain an API Key. Replace the API Key code snippet with your API Key. ChatGPT used to offer free credits to use, but it now costs five or more dollars to purchase API usage.

Reviewer: ```suggestion
3. Create an OpenAI account and follow these [instructions](https://help.openai.com/en/articles/4936850-where-do-i-find-my-api-key) to obtain an API Key. On line 161 of `GPTChatbot.ipynb`, replace the API Key code snippet with your API Key. ChatGPT used to offer free credits to use, but it now costs five or more dollars to purchase API usage.
```

{% img blog/okta-documentation-chatbot/openai-api-key.jpeg alt:"Image with an arrow pointing to where the OpenAI API key should be added in the code." height:"400" %}{: .center-image }

> 4. Download the Okta Developer Documentation files from [here](https://github.com/oktadev/okta-python-chatbot-example/tree/tar-file) and make sure the files are saved in a 

Reviewer: ```suggestion
4. Download the Okta Developer Documentation files from [here](https://github.com/oktadev/okta-python-chatbot-example/tree/tar-file), and extract the archive. Upload the archive contents to a folder named 'oktanaut' in your Google Drive.  
```

>  separate folder on your drive (i.e. not within any folders) and that you are mounting the files from your Google account. Also, make sure to name the folder 'oktanaut'. Alternatively, you can download the files directly from the Okta Developer Documentation Repository. 

Reviewer: ```suggestion
 Make sure the training files are in the `oktanaut` folder, not in an additional folder within it. If you change the name of the folder where you keep the files in your Google Drive, update line 167 of `GPTChatbot.ipynb` to the correct path.  Alternatively, you can download the files directly from the Okta Developer Documentation Repository, or supply your own custom training data.
```

> 5. Run the GPTChatbot.ipynb notebook in your Colab development environment. 

Reviewer: ```suggestion
5. Run the `GPTChatbot.ipynb` notebook in your Colab development environment. 
```

6. After running the script, a web interface displays an input field and a "Chat with Oktanaut!" button.

7. Enter your questions about the Okta Developer Documentation in the input field.

8. To submit your question, click the "Chat with Oktanaut!" button to begin a conversation with the Python chatbot.

9. Oktanaut will answer your question with an AI-generated response using its knowledge from the developer documentation, training data from internal support engineers, and the OpenAI API. The responses will appear on the web interface in real-time.

> 10. Continue the conversation by entering additional questions or prompts and clicking the button. This will lead to better answers due to the self-learning nature of the chatbot.

Reviewer: ```suggestion
10. Continue the conversation by entering additional questions or prompts and clicking the button. The bot will use your questions from earlier in the conversation to improve its understanding of your needs, and provide contextually appropriate answers as you converse with it.
```

> 11. To end the session, say "Thank you," which is also mentioned by the chatbot in its introductory message.

Reviewer: ```suggestion
11. To end the session, say "Thank you" or leave the web page. 
```

{% img blog/okta-documentation-chatbot/openai-oidc.jpeg alt:"Image of Oktanaut's answer to, 'What is OIDC?'" height:"400" %}{: .center-image }

## Further Python Chatbot Improvements
I am enthusiastic about enhancing Oktanaut's training by incorporating information from the Okta Dev Forum and Okta's Software Developer Kits (SDKs). In the future, I hope to collaborate with the Okta Developer Documentation team to improve information gaps in the documentation. I also want to add a feature to automatically update to the latest version of the Okta Developer Documentation so that the data Oktanaut references is up-to-date and reliable.

> Got questions? You can leave them in the comments below! Want to stay in touch? Follow our social channels @oktadev on [Twitter](https://twitter.com/oktadev) and subscribe to our [YouTube](https://www.youtube.com/c/oktadev) channel.

Reviewer: ```suggestion
Have you thought about building your own chatbot with AI? Would you like to know more about how I built Oktanaut? Let me know in the comments below! Want to stay in touch? Follow our social channels @oktadev on [Twitter](https://twitter.com/oktadev) and subscribe to our [YouTube](https://www.youtube.com/c/oktadev) channel.
```

Author: @edunham love the CTA!

---
layout: blog_post
title: "How to Build an Okta Documentation Chatbot in Python"
author: tanish-kumar
by: internal-contributor
communities: [python]

> description: ""

Reviewer 2: @isemona the description is for the preview. You'll want to fill this out.

> tags: []

Reviewer 2: Please add tags. it helps for searching

tweets:
- ""
- ""
- ""
image: blog/okta-documentation-chatbot/oktanaut.jpeg

> type: 

Reviewer 2: Missing

---

> In today's fast-paced world of technology, developer support is essential to ensure a seamless experience for users navigating intricate platforms and APIs. As a Developer Support Intern at Okta, I embarked on a challenging yet rewarding endeavor to create a solution that would streamline accessing and comprehending the Okta developer documentation. The result of my efforts is ✨Oktanaut✨, a versatile Python chatbot designed to assist users in harnessing the power of Okta's developer documentation, available at developer.okta.com. 

Reviewer 2: ```suggestion
In today's fast-paced world of technology, developer support is essential to ensure a seamless experience for users navigating intricate platforms and APIs. As a Developer Support Intern at Okta, I embarked on a challenging yet rewarding endeavor to create a solution that would streamline accessing and comprehending the Okta developer documentation. The result of my efforts is ✨Oktanaut✨, a versatile Python chatbot designed to assist users in harnessing the power of Okta's developer documentation, available at [developer.okta.com](https://developer.okta.com). 
```

> {% img blog/okta-documentation-chatbot/oktadevforumsdk.jpeg alt:"oktadevforumsdk" height:"400" %}{: .center-image } 

Reviewer 2: ```suggestion
{% img blog/okta-documentation-chatbot/oktadevforumsdk.jpeg alt:"OktaDev Forum and SDK logo" height:"400" %}{: .center-image } 
```

Alt tags should be legible and descriptive

– Image source: <cite>[Okta Developer GitHub](https://raw.githubusercontent.com/oktadev/.github/main/images/okta-dev-header.png)</cite> 

> ## Navigating Uncharted Waters

Reviewer 2: Headers are sentence case

This can include keywords relatively easily i think.

> Learning to build a chatbot was a formidable challenge. I spent hours scouring Medium articles, sifting through Coursera courses, and experimenting with various libraries and APIs. Determined, I delved into the world of chatbot development, driven by a desire to create a valuable tool for the Okta community.

Reviewer 2: Should these brands have links to them? I'm almost wondering if this whole paragraph should be reconsidered, IMHO. I'm not sure it adds to the post

## Evolution of a Python Chatbot

> My journey led me to develop two distinct versions of Oktanaut a Python chatbot. The first version harnessed the capabilities of the GPT-3.5 API, allowing it to generate responses based on a broad spectrum of user inquiries. This version could handle various questions, providing a dynamic interaction experience.

Reviewer 2: ```suggestion
My journey led me to develop two distinct versions of "Oktanaut," a Python chatbot. The first version harnessed the capabilities of the GPT-3.5 API, allowing it to generate responses based on a broad spectrum of user inquiries. This version could handle various questions, providing a dynamic interaction experience.
```

Not sure if edit is correct??? something is wrong here

> {% img blog/okta-documentation-chatbot/api.jpeg alt:"api-image" height:"400" %}{: .center-image }

Reviewer 2: Please change alt tags to be human legible and descriptive

– Image source: <cite>[OpenAI APIs with Python — Complete Guide](https://medium.com/@marc.bolle/openai-apis-with-python-complete-guide-d933fb770f95)</cite>

> The second version took a different approach. Using LLamaIndex, I crafted a custom chatbot exclusively trained on the Okta developer documentation. While this version's responses were more accurate due to its focused training, it needed to be more comprehensive in handling a diverse range of questions compared to the GPT-3.5-powered counterpart.

Reviewer 2: Please make sure to link LlamaIndex, and is the capitalization of Okta developer documentation correct for our brand? (idk)

Author: so you are suggesting changing Okta developer documentation to Okta Developer Documentation?

Reviewer: I think the suggestion is to check the style guides linked from https://docs.google.com/document/d/1GjvQQoJJLy4bJIS9Yk5vyzy9tnxDcKcMJ1-yqc-Z-f8/edit. https://github.com/okta/okta-developer-docs/wiki/Style-Guide demonstrates that we do capitalize Dev Docs.

Reviewer: (why yes i did just search our google drive for "style guide" and grab the first promising-looking link there lol)
